x/dclupgrade/keeper/msg_server_propose_upgrade.go

```diff
@@ -0,0 +1,57 @@
+package keeper
+
+import (
+	"context"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	"github.com/zigbee-alliance/distributed-compliance-ledger/x/dclupgrade/types"
+)
+
+func (k msgServer) ProposeUpgrade(goCtx context.Context, msg *types.MsgProposeUpgrade) (*types.MsgProposeUpgradeResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	creatorAddr, err := sdk.AccAddressFromBech32(msg.Creator)
+	if err != nil {
+		return nil, sdkerrors.Wrapf(sdkerrors.ErrInvalidAddress, "Invalid Address: (%s)", err)
+	}
+
+	// check if message creator has enough rights to propose upgrade
+	if !k.dclauthKeeper.HasRole(ctx, creatorAddr, types.UpgradeApprovalRole) {
+		return nil, sdkerrors.Wrapf(sdkerrors.ErrUnauthorized,
+			"MsgProposeUpgrade transaction should be signed by an account with the %s role",
+			types.UpgradeApprovalRole,
+		)
+	}
+
+	// check if proposed upgrade exists
+	_, isFound := k.GetProposedUpgrade(ctx, msg.Plan.Name)
```

Comment:
What if we had already approved and scheduled upgrade, and then an UpgradeProposal is sent for the same name?

Reply:
We allow to create a new proposal in this case. We use the name as the unique identifier of an upgrade proposal, but as to x/upgrade module, it does not prohibit to re-schedule an upgrade with the same name multiple times.

However, in next PRs, when we introduce `ApprovedUpgrade` entity, proposing an upgrade with a previously used name should be prohibited, because we will use the name as the unique identifier of any (proposed or approved) upgrade.